cowbird/permissions_synchronizer.py

```diff
@@ -74,59 +74,178 @@ def __init__(self,
+        # MULTI_TOKEN name tokens can match the related type 0 to N times. They are ignored from the match length since
+        # it is ambiguous in the case of a 0 length match.
+        # SINGLE_TOKEN name tokens can match exactly 1 time only. They are ignored from the match length since we favor
+        # a match with specific names over another with SINGLE_TOKEN generic tokens.
+        matched_res_dict = {}
+        for res_key, res_segments in service_resources.items():
+            match_len = 0
+            res_regex = "^"
+            for segment in res_segments:
+                if segment["name"] == SINGLE_TOKEN:
+                    # match any name with specific type 1 time only
+                    res_regex += rf"/\w+:{segment['type']}"
+                elif segment["name"] == MULTI_TOKEN:
+                    # match any name with specific type, 0 or more times
+                    res_regex += rf"(/\w+:{segment['type']})*"
+                else:
+                    # match name and type exactly
+                    res_regex += f"/{segment['name']:}:{segment['type']}"
+                    match_len += 1
+            res_regex += "$"
+            if re.match(res_regex, resource_nametype_path):
+                matched_res_dict[res_key] = match_len
+
+        # Find the longest match
+        max_matching_keys = [res for res, match_len in matched_res_dict.items()
+                             if match_len == max(matched_res_dict.values())]
+        if len(max_matching_keys) == 1:
+            src_res_key = max_matching_keys[0]
+        elif len(max_matching_keys) > 1:
+            raise ValueError("Found 2 matching resources of the same length in the config. Ambiguous config resources :"
+                             f" {max_matching_keys}")
+        else:
+            raise ValueError("No matching resources could be found in the config file.")
+        return src_res_key, matched_res_dict[src_res_key]
+
+    @staticmethod
+    def _create_res_data(target_segments, src_resource_suffix):
+        # type: (List[Dict], List[Dict]) -> List[Dict]
+        """
+        Creates resource data used to update permissions. This data includes the name and type of each segments of
+        a full resource path.
+
+        @param target_segments List containing the name and type info of each segment of the target resource path.
+        @param src_resource_suffix List similar to the `target_res` argument, but for the input source resource. This
+                                   list contains only the suffix section of the resource path, which is the part that is
+                                   common to both source and target resource paths.
+        """
+        permissions_data = []
+        suffix_target_segments = []
+        # First add 'named' resource data
+        for i, segment in enumerate(target_segments):
+            if segment["name"] in [SINGLE_TOKEN, MULTI_TOKEN]:  # pylint: disable=no-else-break
```

Comment:
warning by pylint is valid here, no need for `else` always executed since `break` did not happen